File: backend/lib/entities/order.rb
class Order
  attr_reader :customer, :payment, :address, :closed_at, :amount

  def initialize(customer)
    @customer = customer
    @address = customer.address
    @grouped_items = {}
    @amount = 0
  end

  def add_product(product, quantity=1)
    item = OrderItem.new(order: self, product: product, quantity: quantity)

    @grouped_items[product.type] ||= []

    @grouped_items[product.type] << item
    @amount += item.total_price
  end

  def items
    @grouped_items.values.flatten
  end

  def items_by_type(type)
    @grouped_items[type]
  end

  def types
    @grouped_items.keys
  end

  def close(closed_at = Time.now)
    @closed_at = closed_at
  end

  def closed?
    !! @closed_at
  end
end

File: backend/Solucao Creditas Challenge/Application/order.py
import time
from customer import Address
from customer import Customer
from product import Product


class OrderItem():
    order = None
    price = 0.0

    def __init__(self, order, productType, price):
        self.order = order
        self.productType = productType
        self.price = price


class OrderTotal(Customer):
    closed_at = None
    items = []
    totalPrice = 0.0
    numOfItems = None

    def __init__(self, customer):
        self.customer = customer.customerName
        self.items = []
        self.numOfItems = None
        self.totalPrice = 0.0

    def putOnCart(self, order_item):
        self.order_item = order_item
        self.items.append([order_item.order, order_item.productType])
        self.numOfItems = len(self.items)
        self.totalPrice += order_item.price

    def close(self, closed_at=time.time()):
        self.closed_at = closed_at
        Product(self.items, self.customer)
